src/python/pants/testutil/rule_runner.py

```diff
@@ -656,89 +659,163 @@ class MockGet(Generic[_O]):
-    category: Literal["Get"] | Literal["Effect"] | str
-    output_type: str | None
-    input_types: tuple[str, ...]
+    category: Literal["Get", "Effect"]
+    output_type: type | None
+    input_types: tuple[type, ...]
 
     def __str__(self):
-        inputs = ",".join(str(e) for e in self.input_types)
-        return f"{self.category}({self.output_type}, ({inputs},))"
+        inputs = ", ".join(e.__name__ for e in self.input_types)
+        return f"{self.category}({self.output_type.__name__}, ({inputs}))"
 
 
 def _compare_expected_mocks(
-    expected: Sequence[Get | Effect | AwaitableConstraints],
-    actual: Sequence[MockGet | MockEffect],
-) -> str:
-    """Try to be helpful with identifying the problem with supplied mocks."""
+    expected: Sequence[AwaitableConstraints],
+    actual_calls: Mapping[str, Callable[..., _O]],
+    actual_gets: Sequence[MockGet | MockEffect],
+    union_membership: UnionMembership | None,
+) -> tuple[str, str]:
+    """Try to be helpful with identifying the problem with supplied mocks.
+
+    Returns a string with errors (or empty if none), and a string with warnings (or empty if none).
+    """
+
+    expected_calls: list[AwaitableConstraints] = []
+    expected_gets: list[AwaitableConstraints] = []
+    for e in expected:
+        (expected_calls if e.rule_id else expected_gets).append(e)
+
+    expected_call_names = {str(e.rule_id) for e in expected_calls}
+    actual_call_names = set(actual_calls.keys())
+
+    missing_call_names = expected_call_names - actual_call_names
+    surplus_call_names = actual_call_names - expected_call_names
 
     def as_comparable(
-        o: Get | Effect | MockGet | MockEffect | AwaitableConstraints | type,
+        o: MockGet | MockEffect | AwaitableConstraints,
     ) -> MockRequestExceptionComparable:
-        if isinstance(o, MockGet) or isinstance(o, Get):
+        category: Literal["Get", "Effect"]
+
+        if isinstance(o, MockGet):
             category = "Get"
-        elif isinstance(o, MockEffect) or isinstance(o, Effect):
+        elif isinstance(o, MockEffect):
             category = "Effect"
         elif isinstance(o, AwaitableConstraints):
             category = "Effect" if o.is_effect else "Get"
         else:
-            # If we don't know of this class, try to give something meaningful
-            # This happened with AwaitableContraints
-            maybe_output_type = getattr(o, "output_type", None)
-            maybe_input_types = getattr(o, "input_types", None)
-            return MockRequestExceptionComparable(
-                category=f"Uncategorised of type {type(o).__qualname__}",
-                output_type=maybe_output_type.__name__ if maybe_output_type is not None else None,
-                input_types=(
-                    tuple(e.__name__ for e in maybe_input_types)
-                    if maybe_input_types is not None
-                    else tuple()
-                ),
+            raise ValueError(
```

Comment:
This is just to double-check our assumption about the possible types of `o` (as referenced in its type annotation), because the previous code had some confusing attempt to handle other types. But this should never trigger.